typechecker/src/checker.rs

```diff
@@ -456,8 +470,10 @@ impl<'a> TraversalVisitor for TypeChecker<'a> {
     }
 
     fn visit_bin_op(&mut self, b: &BinOp) {
-        let l_type = self.infer_expr_type(&b.left);
-        let r_type = self.infer_expr_type(&b.right);
+        // let l_type = self.infer_expr_type(&b.left);
+        self.visit_expr(&b.left);
```

Comment:
I believe this was a bug in the original code -- it mean that the subexpressions don't get their types annotated